extras/indexing/src/main/java/org/apache/rya/api/client/mongo/MongoCreatePCJ.java

```diff
@@ -0,0 +1,103 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.rya.api.client.mongo;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.rya.api.client.CreatePCJ;
+import org.apache.rya.api.client.InstanceDoesNotExistException;
+import org.apache.rya.api.client.InstanceExists;
+import org.apache.rya.api.client.RyaClientException;
+import org.apache.rya.indexing.pcj.storage.PrecomputedJoinStorage.PCJStorageException;
+import org.apache.rya.indexing.pcj.storage.mongo.MongoPcjStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import com.mongodb.MongoClient;
+import com.mongodb.MongoCredential;
+import com.mongodb.ServerAddress;
+
+/**
+ * A Mongo implementation of {@link CreatePCJ}.
+ */
+public class MongoCreatePCJ implements CreatePCJ {
+    private static final Logger log = LoggerFactory.getLogger(MongoCreatePCJ.class);
+
+    private final MongoConnectionDetails connectionDetails;
+    private final InstanceExists instanceExists;
+
+    /**
+     * Constructs an instance of {@link MongoCreatePCJ}.
+     *
+     * @param connectionDetails - Details to connect to the server. (not null)
+     * @param instanceExists - The interactor used to check if a Rya instance exists. (not null)
+     */
+    public MongoCreatePCJ(
+            final MongoConnectionDetails connectionDetails,
+            final MongoInstanceExists instanceExists) {
+        this.connectionDetails = requireNonNull(connectionDetails);
+        this.instanceExists = requireNonNull(instanceExists);
+    }
+
+    @Override
+    public String createPCJ(final String ryaInstanceName, final String sparql, final Set<ExportStrategy> strategies) throws InstanceDoesNotExistException, RyaClientException {
+        requireNonNull(ryaInstanceName);
+        requireNonNull(sparql);
+
+        // Ensure the Rya Instance exists.
+        if (!instanceExists.exists(ryaInstanceName)) {
+            throw new InstanceDoesNotExistException(String.format("There is no Rya instance named '%s'.", ryaInstanceName));
+        }
+
+        final Optional<String> username = connectionDetails.getUsername();
+        final MongoCredential cred;
+        final MongoClient client;
+
+        final ServerAddress addr = new ServerAddress(connectionDetails.getHostname(), connectionDetails.getPort());
```

Comment:
The mongo client this uses should be provided to the interactor instead of having it create a new one. If creating the client is the only thing the connection details are being used for, then you could remove them from the constructor and replace them with a MongoClient.

Reply:
done